pkg/controller/controller.go

```diff
@@ -0,0 +1,242 @@
+		AddFunc: controller.enqueueIAMIdentityMapping,
+		UpdateFunc: func(old, new interface{}) {
+			controller.enqueueIAMIdentityMapping(new)
+		},
+	})
+
+	err := iamMappingInformer.Informer().GetIndexer().AddIndexers(cache.Indexers{
+		"canonicalARN": IndexIAMIdentityMappingByCanonicalArn,
+	})
+	if err != nil {
+		logrus.WithError(err).Fatal("error adding index")
+	}
+
+	controller.iamMappingsIndex = iamMappingInformer.Informer().GetIndexer()
+
+	return controller
+}
+
+// Run will implement the loop for processing items
+func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	logrus.Info("starting aws iam authenticator controller")
+
+	logrus.Info("waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.iamMappingsSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	logrus.Info("starting workers")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	logrus.Info("started workers")
+	<-stopCh
+	logrus.Info("shutting down workers")
+
+	return nil
+}
+
+// runWorker loops over each item looking for boolean returns
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem will process each item off the queue
+func (c *Controller) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			c.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+
+		if err := c.syncHandler(key); err != nil {
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing %s : %s, requeuing", key, err.Error())
+		}
+
+		c.workqueue.Forget(obj)
+		logrus.Infof("successfully synced %s", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+func (c *Controller) syncHandler(key string) (err error) {
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key %s", key))
+		return nil
+	}
+
+	iamIdentityMapping, err := c.iamMappingLister.Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("iam identity mapping %s no longer exists", key))
+			return nil
+		}
+		return err
+	}
+
+	// Process items
+	if iamIdentityMapping.Spec.ARN != "" {
+		iamIdentityMappingCopy := iamIdentityMapping.DeepCopy()
+
+		canonicalizedARN, err := arn.Canonicalize(strings.ToLower(iamIdentityMapping.Spec.ARN))
+		if err != nil {
+			return err
+		}
+
+		iamIdentityMappingCopy.Status.CanonicalARN = canonicalizedARN
+		_, err = c.iamclientset.IamauthenticatorV1alpha1().IAMIdentityMappings().Update(iamIdentityMappingCopy)
```

Comment:
The CRD should have a status subresource and this `UpdateStatus` instead of `Update`. But the status subresource isn't beta until 1.14 so unsure if it's a good idea to make that change now...

Reply:
also, using an in-memory map like before could prevent having to make this API call...but I see that it would be harder to maintain that w/ object updates/deletions.

Reply:
I used `UpdateStatus` in the App Mesh Controller: https://github.com/aws/aws-app-mesh-controller-for-k8s/blob/e10f4585e0fb13b7bc85ffe1b09d2321ad71b338/pkg/controller/virtualnode.go#L153, I don't think I realized it wasn't beta.  It would be good to know if anything is changing as it goes to beta. If not let's use it?

Reply:
I misread docs, it is beta in 1.11 https://github.com/kubernetes/apiextensions-apiserver/blob/96a23c619d2b365bcfd4ccc2fd9f3598105c556f/pkg/features/kube_features.go#L46 . So yeah it can be used